### PR TITLE
Enhance image vector APIs to support multi arch images

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area delivery
/kind enhancement

**What this PR does / why we need it**:
This PR enhances the image vector to support specifying CPU architecture supported by images. If the image doesn't have an architecture specified then it is considered a multi-arch image. The supported architecture types are `amd64`, `arm64`, and `*`, where `*` specifies that the image is multi-arch.

**Which issue(s) this PR fixes**:
Part of gardener/backlog#19

**Special notes for your reviewer**:

**Release note**:

```feature operator
Images in image vector now support a new field `architectures`. It is a list of CPU architecture of machines on which one image can be used. If not specified images are considered to support both `amd64` and `arm64` CPU architecture.
```

### CODE REVIEW

File: pkg/utils/imagevector/types.go [hunk 1]
```diff
@@ -25,6 +34,7 @@ type ImageSource struct {
 	Name           string  `json:"name" yaml:"name"`
 	RuntimeVersion *string `json:"runtimeVersion,omitempty" yaml:"runtimeVersion,omitempty"`
 	TargetVersion  *string `json:"targetVersion,omitempty" yaml:"targetVersion,omitempty"`
+	Architecture   *string `json:"architecture,omitempty" yaml:"architecture,omitempty"`
```
Comment (ialidzhikov) [inline pkg/utils/imagevector/types.go]: Can you elaborate why it is required to make the imagevector pkg architecture aware? Wasn't the idea to have multi-arch images in `charts/images.yaml`? Or did I get it wrong?
Doesn't this change also result in changes in `charts/images.yaml`? As part of this PR I don't see that we specify the new architecture field for any of the images.

File: pkg/apis/core/types_shoot.go [hunk 2]
```diff
@@ -984,6 +984,8 @@ type Worker struct {
 	Zones []string
 	// MachineControllerManagerSettings contains configurations for different worker-pools. Eg. MachineDrainTimeout, MachineHealthTimeout.
 	MachineControllerManagerSettings *MachineControllerManagerSettings
+	// Architecture is CPU architecture of machines in this worker pool.
+	Architecture *string
```
Comment (ialidzhikov) [inline pkg/apis/core/types_shoot.go]: According to the last discussions such field would make sense only if a machine type can support multiple architectures. Did you double-checked this?

File: pkg/apis/core/types_shoot.go [hunk 3]
```diff
@@ -984,6 +984,8 @@ type Worker struct {
 	Zones []string
 	// MachineControllerManagerSettings contains configurations for different worker-pools. Eg. MachineDrainTimeout, MachineHealthTimeout.
 	MachineControllerManagerSettings *MachineControllerManagerSettings
+	// Architecture is CPU architecture of machines in this worker pool.
```
Comment (ialidzhikov) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// Architecture is the CPU architecture of the machines in this worker pool.
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 4]
```diff
@@ -1233,6 +1233,9 @@ type Worker struct {
 	// MachineControllerManagerSettings contains configurations for different worker-pools. Eg. MachineDrainTimeout, MachineHealthTimeout.
 	// +optional
 	MachineControllerManagerSettings *MachineControllerManagerSettings `json:"machineControllerManager,omitempty" protobuf:"bytes,19,opt,name=machineControllerManager"`
+	// Architecture is CPU architecture of machines in this worker pool.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/types_shoot.go]:
```suggestion
	// Architecture is the CPU architecture of the machines in this worker pool.
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 5]
```diff
@@ -1245,6 +1245,9 @@ type Worker struct {
 	// MachineControllerManagerSettings contains configurations for different worker-pools. Eg. MachineDrainTimeout, MachineHealthTimeout.
 	// +optional
 	MachineControllerManagerSettings *MachineControllerManagerSettings `json:"machineControllerManager,omitempty" protobuf:"bytes,19,opt,name=machineControllerManager"`
+	// Architecture is CPU architecture of machines in this worker pool.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// Architecture is the CPU architecture of the machines in this worker pool.
```

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (ialidzhikov) [inline pkg/apis/core/types_shoot.go]: Wouldn't it be better to place this field under the machine field?

```yaml
    workers:
    - name: cpu-worker
      minimum: 3
      maximum: 5
      machine:
        type: m5.large
        architecture: amd64
        image:
          name: <some-image-name>
          version: <some-image-version>
```

cc @rfranzke

File: pkg/apis/core/validation/shoot.go [hunk 6]
```diff
@@ -1702,6 +1707,15 @@ func ValidateContainerRuntimes(containerRuntime []core.ContainerRuntime, fldPath
 	return allErrs
 }
 
+func ValidateArchitecture(arch *string, fldPath *field.Path) field.ErrorList {
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go]: Missing doc string.

File: pkg/apis/core/validation/shoot.go [hunk 7]
```diff
@@ -1702,6 +1707,15 @@ func ValidateContainerRuntimes(containerRuntime []core.ContainerRuntime, fldPath
...
+	allErrs := field.ErrorList{}
+	if arch != nil && *arch != imagevector.X86 && *arch != imagevector.Arm64 {
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go]: Can we define new constants instead of reusing the ones from the imagevector pkg?

File: pkg/apis/core/validation/shoot_test.go [hunk 8]
```diff
@@ -4282,6 +4283,22 @@ var _ = Describe("Shoot Validation Tests", func() {
...
 		)
 
+		DescribeTable("validate architecture",
+			func(arch *string, matcher gomegatypes.GomegaMatcher) {
+				errList := ValidateArchitecture(arch, field.NewPath("architecture"))
+
+				Expect(errList).To(matcher)
+			},
+
+			Entry("<empty> is a valid architecture name", nil, HaveLen(0)),
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
			Entry("nil is a valid architecture name", nil, HaveLen(0)),
```

**Review by ialidzhikov — COMMENTED** [review]
Plus, the PR title and the PR description (currently empty) do no reveal anything about what this PR aims to tackle. Can you provide meaningful PR title, description and release note?

File: pkg/utils/imagevector/types.go [hunk 9]
```diff
@@ -14,6 +14,15 @@
 
 package imagevector
 
+const (
+	// X86 is a constant for the 'x86' architecture.
+	X86 = "x86"
+	// Arm64 is a constant for the 'arm64' architecture.
+	Arm64 = "arm64"
+	// WildcardArchitecture is a constant for multi architectures.
+	WildcardArchitecture = "*"
```
Comment (ialidzhikov) [inline pkg/utils/imagevector/types.go]: K8s maintains the architecture in the Node status and as a Node label. It uses the `goruntime.GOARCH` call to determine this field:

https://github.com/kubernetes/kubernetes/blob/23144130a815117194afbf3f703024ed76d5dac1/pkg/kubelet/nodestatus/setters.go#L440-L446


```yaml
kind: Node
# ...
status:
  nodeInfo:
    architecture: amd64
    # ...
    operatingSystem: linux
    osImage: Garden Linux 576.9
```


I don't see `x86` being a valid value for `goruntime.GOARCH` - ref https://gist.github.com/asukakenji/f15ba7e588ac42795f421b48b8aede63#goarch-values. The rather use `amd64` and `amd64p32`.

I don't have much expertise in the topic but shouldn't we stick to what K8s uses and maintains as allowed values?

File: pkg/utils/imagevector/types.go (hunk 1 above)
Comment (acumino) [reply pkg/utils/imagevector/types.go]: > Can you elaborate why it is required to make the imagevector pkg architecture aware? Wasn't the idea to have multi-arch images in charts/images.yaml? Or did I get it wrong?

Our goal is to have as many possible multi arch images but there is a chance we can't achieve it. As discussed with @rfranzke we need to add a new field.

```
I think it’s questionable whether we can achieve having only multi-arch images since there are some
images that we rely on which are out of our control (if community does not help in timely manner then what?). 
But even if we make it, what if we introduce a new image tomorrow where only separate images are available and community does not want to build multi-arch image for whatever reasons? In this case, we have an issue with ARM shoots
```
> Doesn't this change also result in changes in charts/images.yaml? As part of this PR I don't see that we specify the new architecture field for any of the images.

Yes, the changes in `charts/images.yaml` are required. If an image doesn't have arch filed it is by default considered as amd64 based image. I will update the `image.yaml` once the PR is ready for review based on what are images available at that time.

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (acumino) [reply pkg/apis/core/types_shoot.go]: Machines can't support multiple architectures. But we still need it to select the images for the shoot based on the shoot CPU architecture.

Comment (acumino) [reply pkg/apis/core/types_shoot.go]: Since one worker group has same type of machines I guess it is better to have it as part of the worker. But other approach also looks good.

Comment (rfranzke) [reply pkg/apis/core/types_shoot.go]: I'm fine with both, slightly prefer `machine.architecture` as well but no strong opinion

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 10]
```diff
@@ -663,6 +663,11 @@ const (
 	DNSRecordExternalName = "external"
 	// DNSRecordOwnerName is a constant for DNSRecord objects used for the owner domain name.
 	DNSRecordOwnerName = "owner"
+
+	// Amd64 is a constant for the 'amd64' architecture.
+	Amd64 = "amd64"
+	// Arm64 is a constant for the 'arm64' architecture.
+	Arm64 = "arm64"
```
Comment (timebertt) [inline pkg/apis/core/v1beta1/constants/types_constants.go]: This is a common change with https://github.com/gardener/gardener/pull/6178.
Which one do you want to get merged first?
Please mark the other one as draft accordingly.

File: pkg/apis/core/validation/shoot.go [hunk 11]
```diff
@@ -1242,6 +1242,10 @@ func ValidateWorker(worker core.Worker, kubernetesVersion string, fldPath *field
 		allErrs = append(allErrs, ValidateCRI(worker.CRI, kubernetesVersion, fldPath.Child("cri"))...)
 	}
 
+	if worker.Machine.Architecture != nil {
```
Comment (timebertt) [inline pkg/apis/core/validation/shoot.go]: Again, defaulting should be in place to default to the current implicit value of `amd64`

File: pkg/apis/core/validation/shoot.go [hunk 12]
```diff
@@ -1696,6 +1700,16 @@ func ValidateContainerRuntimes(containerRuntime []core.ContainerRuntime, fldPath
 	return allErrs
 }
 
+// ValidateArchitecture validates the CPU architecure of the machines in this worker pool.
+func ValidateArchitecture(arch *string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if arch != nil && *arch != v1beta1constants.Amd64 && *arch != v1beta1constants.Arm64 {
+		allErrs = append(allErrs, field.Invalid(fldPath, arch, fmt.Sprintf("specified architecture is invalid %s", *arch)))
+	}
```
Comment (timebertt) [inline pkg/apis/core/validation/shoot.go]: Same as https://github.com/gardener/gardener/pull/6178#discussion_r907219860

File: pkg/utils/imagevector/imagevector.go [hunk 13]
```diff
@@ -180,7 +191,12 @@ func (o *FindOptions) String() string {
 		targetVersion = "target version " + *o.TargetVersion
 	}
 
-	return runtimeVersion + targetVersion
+	var architecture string
+	if o.Architecture != nil {
+		architecture = "architecture " + *o.Architecture + " "
```
Comment (timebertt) [inline pkg/utils/imagevector/imagevector.go]: Spacing in the resulting string looks off

File: pkg/utils/imagevector/types.go [hunk 14]
```diff
@@ -14,6 +14,15 @@
 
 package imagevector
 
+const (
+	// Amd64 is a constant for the 'amd64' architecture.
+	Amd64 = "amd64"
+	// Arm64 is a constant for the 'arm64' architecture.
+	Arm64 = "arm64"
+	// WildcardArchitecture is a constant for multi architectures.
+	WildcardArchitecture = "*"
```
Comment (timebertt) [inline pkg/utils/imagevector/types.go]: can we reuse constants in `v1beta1constants` package?

File: pkg/apis/core/types_shoot.go [hunk 15]
```diff
@@ -1022,6 +1022,8 @@ type WorkerKubernetes struct {
 type Machine struct {
 	// Type is the machine type of the worker group.
 	Type string
+	// Architecture is the CPU architecture of the machines in this worker pool.
+	Architecture *string
```
Comment (timebertt) [inline pkg/apis/core/types_shoot.go]: shoot example in `example` should be adapted.

File: pkg/utils/imagevector/types.go (hunk 1 above)
Comment (timebertt) [inline pkg/utils/imagevector/types.go]: Can you document this change in `docs/deployment/image_vector.md`

File: pkg/utils/imagevector/types.go (hunk 14 above)
Comment (acumino) [reply pkg/utils/imagevector/types.go]: We should not, It will make imagevector package dependent on apis.

Comment (timebertt) [reply pkg/utils/imagevector/types.go]: And why would that be a bad thing?

Comment (acumino) [reply pkg/utils/imagevector/types.go]: If someone wants to use image vector pkg, apis package will also be required for it, just because of two constants.

Comment (timebertt) [reply pkg/utils/imagevector/types.go]: The image vector pkg is not a proper "exposed" API and it's not supposed to be used as such, that's why it's located under `pkg/utils`. Hence, we shouldn't care about someone importing this package.
Even if we would, that someone would probably import the API packages anyway. And even if not, the API constants package is very small

TL;DR: please use the constants

File: docs/deployment/image_vector.md [hunk 16]
```diff
@@ -13,16 +13,47 @@ images:
...
+  architecture: "*"
+...
+```
+
+Architecture is an optional field of image. It specifes CPU architecture of machines on which this image can be used. There are three valid option for architecture field as follow:
+ - `amd64` : This sepecifies image can run only on machines having CPU architecture `amd64`.
+ - `arm64` : This sepecifies image can run only on machines having CPU architecture `arm64`.
+ - `*` : This sepecifies image can run on machines having CPU architecture either `amd64` or `arm64`.
+
+If image doesn't specifies any architecture then by default it's architecture is considered `*`.
```
Comment (acumino) [inline docs/deployment/image_vector.md]:
```suggestion 
- `amd64`: This specifies image can run only on machines having CPU architecture `amd64`.
 - `arm64`: This specifies image can run only on machines having CPU architecture `arm64`.
 - `*`: This specifies image can run on machines having CPU architecture either `amd64` or `arm64`.

If the image doesn't specify any architecture then by default its architecture is considered `*`.
```

File: docs/deployment/image_vector.md [hunk 17]
```diff
@@ -13,16 +13,47 @@ images:
...
+  repository: k8s.gcr.io/pause
+  tag: "3.0"
+  architecture: "*"
+...
+```
+
+Architecture is an optional field of image. It specifes CPU architecture of machines on which this image can be used. There are three valid option for architecture field as follow:
+- `amd64`: This specifies image can run only on machines having CPU architecture `amd64`.
+ - `arm64`: This specifies image can run only on machines having CPU architecture `arm64`.
+ - `*`: This specifies image can run on machines having CPU architecture either `amd64` or `arm64`.
```
Comment (timebertt) [inline docs/deployment/image_vector.md]:
```suggestion
- `arm64`: This specifies image can run only on machines having CPU architecture `arm64`.
- `*`: This specifies image can run on machines having CPU architecture either `amd64` or `arm64`.
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 18]
```diff
@@ -1279,6 +1279,9 @@ type WorkerKubernetes struct {
 type Machine struct {
 	// Type is the machine type of the worker group.
 	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Architecture is the CPU architecture of the machines in this worker pool.
+	// +optional
+	Architecture *string `json:"architecture,omitempty" protobuf:"bytes,20,opt,name=architecture"`
```
Comment (timebertt) [inline pkg/apis/core/v1alpha1/types_shoot.go]: - new fields new to be added to the end of the struct
- this field needs to be cross-checked against the cloudprofile, whether the machine type and image support the specified archictecture
- so probably move this change to a dedicated PR that adds the validation once https://github.com/gardener/gardener/pull/6178 is merged

File: pkg/utils/imagevector/imagevector.go [hunk 19]
```diff
@@ -227,23 +251,57 @@ func checkConstraint(constraint, version *string) (score int, ok bool, err error
...
+	if source != nil {
+		sourceArch = *source
+	}
+	if desired != nil {
+		desiredArch = *desired
+	}
+
+	if desiredArch == v1beta1constants.ArchitectureWildcard {
+		return 0, false, fmt.Errorf("desired architecture cannot be '*'")
+	}
```
Comment (timebertt) [inline pkg/utils/imagevector/imagevector.go]: can be dropped, as the shoot would already be validated to not specify the wildcard

**Review by timebertt — CHANGES_REQUESTED** [review]
Joint review with @rfranzke @acumino

File: docs/deployment/image_vector.md [hunk 20]
```diff
@@ -13,16 +13,47 @@ images:
   repository: gcr.io/google_containers/pause-amd64
   tag: "3.0"
   version: 1.17.x
+  architecture: "amd64"
```
Comment (rfranzke) [inline docs/deployment/image_vector.md]: As discussed, let's make this 

```yaml
architectures:
- amd64
- arm64
```

This way, we can remove the wildcard constant. The implicit default (if `len(architectures)==0`) should be `[amd64,arm64]`.

**Comment by rfranzke** [comment]
Can you please add a proper release note?

**Comment by acumino** [comment]
> Can you please add a proper release note?

Done, PTAL.

**Review by rfranzke — COMMENTED** [review]
Thanks!
/milestone v1.50
/lgtm
/approve

File: pkg/utils/imagevector/types.go [hunk 21]
```diff
@@ -22,9 +22,10 @@ package imagevector
 // in the seed cluster and act on the shoot cluster. Different versions might be used depending on the
 // seed and the shoot version.
 type ImageSource struct {
-	Name           string  `json:"name" yaml:"name"`
-	RuntimeVersion *string `json:"runtimeVersion,omitempty" yaml:"runtimeVersion,omitempty"`
-	TargetVersion  *string `json:"targetVersion,omitempty" yaml:"targetVersion,omitempty"`
+	Name           string    `json:"name" yaml:"name"`
+	RuntimeVersion *string   `json:"runtimeVersion,omitempty" yaml:"runtimeVersion,omitempty"`
+	TargetVersion  *string   `json:"targetVersion,omitempty" yaml:"targetVersion,omitempty"`
+	Architecture   *[]string `json:"architecture,omitempty" yaml:"architecture,omitempty"`
```
Comment (acumino) [inline pkg/utils/imagevector/types.go]: @rfranzke Architecture field here is fine or do we need to make it as suggested? Without pointer also marking an array as an optional field can work.
```suggestion
	Architecture   []string `json:"architecture,omitempty" yaml:"architecture,omitempty"`
```

Comment (timebertt) [reply pkg/utils/imagevector/types.go]: Yes, slices should not have a pointer.

Comment (rfranzke) [reply pkg/utils/imagevector/types.go]: /approve cancel

I missed this, yes, let's remove the pointer.

Comment (timebertt) [reply pkg/utils/imagevector/types.go]: Also, the field name should be plural

File: docs/deployment/image_vector.md [hunk 22]
```diff
@@ -13,16 +13,52 @@ images:
   repository: gcr.io/google_containers/pause-amd64
   tag: "3.0"
   version: 1.17.x
+  architecture:
```
Comment (rfranzke) [inline docs/deployment/image_vector.md]:
```suggestion
  architectures:
```

same for all occurrences below

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
